Refactor/switch to roles

NichArchA82 opened the pull request:
### **PR Type**
Enhancement


___

### **Description**
- Refactor monolithic playbook into modular Ansible roles

- Split server setup into 6 specialized roles

- Improve code organization and maintainability

- Convert templates to static files where possible


___

### Diagram Walkthrough


```mermaid
flowchart LR
  A["Monolithic Playbook"] --> B["discover_ssh_port"]
  A --> C["setup_server"]
  A --> D["configure_snapshots"]
  A --> E["setup_libvirt"]
  A --> F["setup_docker"]
  A --> G["setup_tailscale"]
  A --> H["setup_cache"]
```



<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody></tr></tbody></table>

</details>

___

NichArchA82 commented:
The dns resolve step as mentioned by ai was removed for a couple reasons.
1. Tailscale manages the DNS files as of right now, so putting that file in would cause it to get overwritten, or everytime we ran this we would overwrite the tailscale config causing discrepancies between servers that have the tailscale config and ones that do not.
2. tailscale integrates with systemd-resolved so that step was removed to remove it. This is done as the issue we were seeing with it was dnsmasq was conflicting with it, but we fixed that by turning DNS off so it seems to be recommended to not remove systemd-resolved and let tailscale integrate with it to manage it.

NichArchA82 commented:
I resolved all the copilot feedback of changing the roles path as I have tested with the roles and the files/file_to_copy, i.e files/htoprc in the setup_server role copies fine and ansible detects it. I am unsure why it was getting confused of relative paths, vs working paths.

venkatamutyala commented:
@NichArchA82 can we keep the resolve conf? Although we generally have tailscale do it's thing this is usually nice to have in case the DNS provider from tailscale is having issues.

If we can do cloudflare, Google, then quad999 that would be preferred. Basically have it match what we do in tailscale except it won't be able to resolve talent hostnames.

venkatamutyala commented:
> or if we have tailscale manage it, we keep the resolv.conf file out and leave the code as is. I see pros and cons to each so I don't have a strong preference either way.

Understood. We can turn tailscale off/on at anytime in the console with a quick button click

venkatamutyala commented:
The goal i have with these NS changes is to have a consistent setup across providers. So hetzner has their own DNS while OVH may have their own resolver and each of their resolvers may behave differently.

NichArchA82 commented:
@venkatamutyala I think with your suggestions and reading more into systemd-resolved, I think it would be best to add the code back to remove systemd-resolvd on our servers. Add the --accept-dns=false flag to the tailscale up command. Manage the /etc/resolv.conf with ansible, we can add tailscale dns as the first name server then fallback to the ones recommended so we still get tailscale DNS, but don't have tailscale managing it or overwriting it, so we have one specific place to edit it. Unless you have any objections. I question the future of systemd-resolvd (it got removed from Trixie repos) due to conflicts between mDNS as [referenced](https://www.reddit.com/r/debian/comments/1jmoq7y/systemdresolved_removed_from_unstable_function/) while I understand we are not using Trixie yet, it seems like a good idea to remove it now as we already did that before. Having this code will also align with ovh, and other clouds that may not be using systemd-resolvd by default.

NichArchA82 commented:
> e LTS

I just saw this. I opted to remove it because it is just a safer bet and you brought up ovh which was one of the reasons we removed it to begin with. During the refactor I was opting to not remove systemd because according to chatgpt and gemini, it is the modern way of dealing with DNS, but managing the /etc/resolv.conf file is the old approach and is perfectly valid from my understanding. However, we need to pick a way of doing it and commit. systemd-resolved manages the /etc/resolv.conf file and you risk that overwriting it. Tailscale also does the same, it can integrate with systemd-resolvd to manage it, but if that package is not available then it will fallback to the /etc/resolv.conf file. None of the servers in prod or stage have the systemd-resolvd package because we ran this ansible on it before I was thinking about adding it back. If we change this now, we will need to go back and install it on all the servers. Not to mention that trixie is concerning. You can see the changes I went for, but I opted to just have ansible manually edit the file rather then having all these external packages that can edit it. This also makes it very hard to manage DNS and we will have configuration drift having three packages (systemd-resolvd, tailscale, ansible) managing the same DNS file entry and figuring out what servers are on what config from which package. Highlighted changes are remove systemd-resolvd like we did before I refactored the code, have ansible manage the /etc/resolv.conf file, (add in the DNS entry for tailscale so magicdns works for fqdn in the tailnet) and in the tailscale up command add --accept-dns-false so tailscale won't manage this file at all.
